modules/openvr/openvr.cpp

```diff
@@ -0,0 +1,381 @@
+
+	return hmd != NULL;
+};
+
+void OpenVR::uninitialize() {
+	if (hmd != NULL) {
+		for (uint32_t i = 0; i < vr::k_unMaxTrackedDeviceCount; i++) {
+			detach_device(i);
+		};
+
+		vr::VR_Shutdown();
+		hmd = NULL;
+		render_models = NULL;
+	};
+};
+
+Point2 OpenVR::get_recommended_render_targetsize() {
+	if (hmd != NULL) {
+		uint32_t width, height;
+
+		hmd->GetRecommendedRenderTargetSize(&width, &height);
+
+		return Point2(width, height);
+	} else {
+		return Point2(512,512);
+	};
+};
+
+Transform OpenVR::get_transform_for_eye(ArVrInterface::Eyes p_eye, const Transform& p_head_position) {
+	Transform newtransform;
+
+	if (hmd != NULL) {
+		vr::HmdMatrix34_t matrix = hmd->GetEyeToHeadTransform(p_eye == ArVrInterface::EYE_LEFT ? vr::Eye_Left : vr::Eye_Right);
+
+		newtransform.basis.set(
+			matrix.m[0][0], matrix.m[0][1], matrix.m[0][2],
+			matrix.m[1][0], matrix.m[1][1], matrix.m[1][2],
+			matrix.m[2][0], matrix.m[2][1], matrix.m[2][2]
+		);
+
+		newtransform.origin.x = matrix.m[0][3];
+		newtransform.origin.y = matrix.m[1][3];
+		newtransform.origin.z = matrix.m[2][3];
+	} else {
+		if (p_eye == ArVrInterface::EYE_LEFT) {
+			newtransform.origin.x = -0.035;
+		} else {
+			newtransform.origin.x = 0.035;
+		};
+	};
+
+	newtransform = p_head_position * newtransform;
+
+	return newtransform;
+};
+
+Rect2 OpenVR::get_frustum_for_eye(ArVrInterface::Eyes p_eye) {
+	Frustum frustum;
+	if (hmd != NULL) {
+		Point2 size = get_recommended_render_targetsize();
+
+		// note that openvr is upside down in relation to godot
+		hmd->GetProjectionRaw(p_eye == ArVrInterface::EYE_LEFT ? vr::Eye_Left : vr::Eye_Right, &frustum.left, &frustum.right, &frustum.bottom, &frustum.top);
+	
+		// Godot will (re)apply our aspect ratio so we need to unapply it
+		frustum.left = frustum.left * size.y / size.x;
+		frustum.right = frustum.right * size.y / size.x;
+	} else {
+		// just return a pretty basic stereoscopic frustum
+		frustum.set_frustum(60.0, p_eye == ArVrInterface::EYE_LEFT ? Frustum::EYE_LEFT : Frustum::EYE_RIGHT, 0.065, 1.0);
+	};
+
+///@TODO Return Frustum instead of Rect2, need to add Frustum support to Variants
+//	return frustum;
+	return Rect2(frustum.left, frustum.top, frustum.right, frustum.bottom);
+};
+
+bool OpenVR::get_need_commit_eyes() {
+	return true;
+};
+
+void OpenVR::commit_eye_texture(ArVrInterface::Eyes p_eye, Node* p_viewport) {
+	///@TODO this needs to be greatly improved, I'm sure it can be sped up...
+
+	if (hmd != NULL) {
+		uint32_t texid;
+		Ref<Texture> eye_texture;
+
+		Viewport *vp=p_viewport->cast_to<Viewport>();
+		ERR_FAIL_COND(!vp);
+
+		Ref<ViewportTexture> rtt = vp->get_texture();
+		eye_texture=rtt;
+
+		// added texture_get_texid to visual server (and rasterers), possibly temporary solution as gles3 may change access to the opengl texture id
+		if (eye_texture.is_valid()) {
+			// upside down!
+			vr::VRTextureBounds_t bounds;
+			bounds.uMin = 0.0;
+			bounds.uMax = 1.0;
+			bounds.vMin = 1.0;
+			bounds.vMax = 0.0;
+
```

Comment:
@Hinsbart Once I'm behind my PC (I'm on my mac) I'll change this but this is the code I was talking about, for Godot 2 I had to flip the viewport upside down, so change this to:
bounds.uMin = 0.0;
bounds.uMax = 1.0;
bounds.vMin = 0.0;
bounds.vMax = 1.0;

and it should no longer render upside down